projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php

```diff
@@ -166,44 +166,80 @@ public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAna
-				$preserved_scripts[ $handle ] = clone $script;
-			}
-		}
+			// Initialize fresh asset registries to control what gets loaded
+			$wp_styles  = new WP_Styles();
+			$wp_scripts = new WP_Scripts();
 
-		foreach ( $current_wp_styles->registered as $handle => $style ) {
-			if ( $this->is_allowed_plugin_handle( $handle ) && ! $this->is_core_or_gutenberg_asset( $style->src ) ) {
-				$preserved_styles[ $handle ] = clone $style;
-			}
-		}
+			// Restore preserved plugin assets
+			$this->restore_preserved_assets( $preserved );
 
-		// Initialize new instances to control which assets are loaded
-		$wp_styles  = new WP_Styles();
-		$wp_scripts = new WP_Scripts();
+			// Set up a block editor screen context to prevent errors when
+			// plugins/themes call get_current_screen() during asset enqueueing
+			$this->setup_block_editor_screen();
 
-		// Restore preserved plugin scripts
-		foreach ( $preserved_scripts as $handle => $script ) {
-			$wp_scripts->registered[ $handle ] = $script;
-		}
+			// Trigger wp_loaded action that plugins frequently use to enqueue assets.
+			// This must happen after screen setup and before we collect enqueued assets.
+			do_action( 'wp_loaded' );
 
-		foreach ( $preserved_styles as $handle => $style ) {
-			$wp_styles->registered[ $handle ] = $style;
-		}
+			// Enqueue all core WordPress editor assets
+			$this->enqueue_core_editor_assets();
+
+			// Trigger block editor asset actions with forced script/style loading
+			add_filter( 'should_load_block_editor_scripts_and_styles', '__return_true' );
+			do_action( 'enqueue_block_assets' );
+			do_action( 'enqueue_block_editor_assets' );
+			remove_filter( 'should_load_block_editor_scripts_and_styles', '__return_true' );
+
+			// Enqueue editor-specific assets for all registered block types
+			$this->enqueue_block_type_editor_assets();
+
+			// Remove disallowed plugin assets before generating output
+			$this->unregister_disallowed_plugin_assets();
 
-		// Set up a block editor screen context to prevent errors when
-		// plugins/themes call get_current_screen() during asset enqueueing
-		$this->setup_block_editor_screen();
+			// Capture HTML output with absolute URLs
+			$html = $this->with_absolute_urls(
+				function () {
+					return array(
+						'styles'  => $this->capture_styles_output(),
+						'scripts' => $this->capture_scripts_output(),
+					);
+				}
+			);
 
-		// Trigger an action frequently used by plugins to enqueue assets.
-		do_action( 'wp_loaded' );
+			return rest_ensure_response(
+				array(
+					'allowed_block_types' => array_merge(
+						$this->get_core_block_types(),
+						self::ALLOWED_PLUGIN_BLOCKS
+					),
+					'scripts'             => $html['scripts'],
+					'styles'              => $html['styles'],
+				)
+			);
+
+		} finally {
+			// Always restore original asset state, even if an exception occurred
+			$wp_styles  = $current_wp_styles;
+			$wp_scripts = $current_wp_scripts;
+		}
```

Comment:
This `try`/`finally` block as added to ensure global state is restored in the event of an error. It will likely have no impact, but felt worthwhile nonetheless.